devito/types/sparse.py

```diff
@@ -931,61 +1035,157 @@ class PrecomputedSparseFunction(AbstractSparseFunction):
 
         # Grid points per sparse point (2 in the case of bilinear and trilinear)
         r = kwargs.get('r')
-        gridpoints = kwargs.get('gridpoints')
-        interpolation_coeffs = kwargs.get('interpolation_coeffs')
+        if not is_integer(r):
+            raise TypeError('Need `r` int argument')
+        if r <= 0:
+            raise ValueError('`r` must be > 0')
+
+        self._radius = r
 
-        self.interpolator = PrecomputedInterpolator(self, r, gridpoints,
-                                                    interpolation_coeffs)
+        coordinates = kwargs.get('coordinates', kwargs.get('coordinates_data'))
+        gridpoints = kwargs.get('gridpoints', kwargs.get('gridpoints_data'))
+        if coordinates is not None and gridpoints is not None:
+            raise ValueError("Either `coordinates` or `gridpoints` must be "
+                             "provided, but not both")
+
+        # Specifying only `npoints` is acceptable; this will require the user
+        # to setup the coordinates data later on
+        npoint = kwargs.get('npoint', None)
+        if self.npoint and coordinates is None and gridpoints is None:
+            coordinates = np.zeros((npoint, self.grid.dim))
+
+        if coordinates is not None:
+            self._coordinates = self.__subfunc_setup__(coordinates, 'coords')
+            self._gridpoints = None
+            self._dist_origin = {self._coordinates: self.grid.origin_offset}
+        else:
+            assert gridpoints is not None
+            self._coordinates = None
+            self._gridpoints = self.__subfunc_setup__(gridpoints, 'gridpoints')
+            self._dist_origin = {self._gridpoints: self.grid.origin_ioffset}
+
+        # Setup the interpolation coefficients. These are compulsory
+        interpolation_coeffs = kwargs.get('interpolation_coeffs',
+                                          kwargs.get('interpolation_coeffs_data'))
+        self._interpolation_coeffs = \
+            self.__subfunc_setup__(interpolation_coeffs, 'interp_coeffs')
+        self._dist_origin.update({self._interpolation_coeffs: None})
+        # Make sure it matches the radius
+        if self._interpolation_coeffs.shape[-1] != r:
+            nr = self._interpolation_coeffs.shape[-1]
+            raise ValueError("Interpolation coefficients shape %d do "
+                             "not match specified radius %d" % (r, nr))
+
+        warning("Ensure that the provided interpolation coefficient and grid "
+                "point values are computed on the final grid that will be used "
+                "for other computations.")
+
+        self.interpolator = PrecomputedInterpolator(self)
+
+    @property
+    def r(self):
+        return self._radius
+
+    @cached_property
+    def _point_increments(self):
+        """Index increments in each dimension for each point symbol."""
+        return tuple(product(range(-self.r//2+1, self.r//2+1), repeat=self.grid.dim))
```

Comment:
considering that `r == _radius`, could we move this property to one of the superclasses?